chronos/streams/tlsstream.nim

```diff
@@ -0,0 +1,334 @@
+#              MIT license (LICENSE-MIT)
+
+## This module implements TLS stream reading and writing.
+import bearssl, bearssl/cacert
+import ../asyncloop, ../timer, ../asyncsync
+import asyncstream, ../transports/stream, ../transports/common
+import strutils
+
+type
+  TLSStreamKind {.pure.} = enum
+    Client, Server
+
+  TLSVersion* {.pure.} = enum
+    TLS10 = 0x0301, TLS11 = 0x0302, TLS12 = 0x0303
+
+  TLSFlags* {.pure.} = enum
+    NoVerifyHost,       # Client: Skip remote certificate check
+    NoVerifySN,         # Client: Skip Server Name Indication (SNI) check
+    NoRenegotiation,    # Server: Reject renegotiations requests
+    NoClientAuth,       # Server: Disable strict client authentication
+    FailOnAlpnMismatch  # Server: Fail on application protocol mismatch
+
+type
+  TlsStreamWriter* = ref object of AsyncStreamWriter
+    case kind: TlsStreamKind
+    of TlsStreamKind.Client:
+      ccontext: ptr SslClientContext
+    of TlsStreamKind.Server:
+      scontext: ptr SslServerContext
+    stream*: TlsAsyncStream
+    switchToReader*: AsyncEvent
+    switchToWriter*: AsyncEvent
+
+  TlsStreamReader* = ref object of AsyncStreamReader
+    case kind: TlsStreamKind
+    of TlsStreamKind.Client:
+      ccontext: ptr SslClientContext
+    of TlsStreamKind.Server:
+      scontext: ptr SslServerContext
+    stream*: TlsAsyncStream
+    switchToReader*: AsyncEvent
+    switchToWriter*: AsyncEvent
+
+  TlsAsyncStream* = ref object of RootRef
+    xwc*: X509NoAnchorContext
+    context*: SslClientContext
+    sbuffer*: seq[byte]
+    x509*: X509MinimalContext
+    reader*: TlsStreamReader
+    writer*: TlsStreamWriter
+
+  TlsStreamError* = object of CatchableError
+  TlsStreamProtocolError* = object of TlsStreamError
+    errCode*: int
+
+template newTlsStreamProtocolError[T](message: T): ref Exception =
+  var msg = ""
+  var code = 0
+  when T is string:
+    msg.add(message)
+  elif T is cint:
+    msg.add(sslErrorMsg(message) & " (code: " & $int(message) & ")")
+    code = int(message)
+  elif T is int:
+    msg.add(sslErrorMsg(message) & " (code: " & $message & ")")
+    code = message
+  else:
+    msg.add("Internal Error")
+  var err = newException(TlsStreamProtocolError, msg)
+  err.errCode = code
+  err
+
+# proc raiseTlsStreamProtoError*[T](message: T) =
+#   raise newTlsStreamProtocolError(message)
+
+proc getStringState*(state: cuint): string =
+  var n = newSeq[string]()
+  if (state and SSL_CLOSED) == SSL_CLOSED:
+    n.add("Closed")
+  if (state and SSL_SENDREC) == SSL_SENDREC:
+    n.add("SendRec")
+  if (state and SSL_RECVREC) == SSL_RECVREC:
+    n.add("RecvRec")
+  if (state and SSL_SENDAPP) == SSL_SENDAPP:
+    n.add("SendApp")
+  if (state and SSL_RECVAPP) == SSL_RECVAPP:
+    n.add("RecvApp")
+  result = "{" & n.join(", ") & "} number (" & $state & ")"
+
+proc tlsWriteLoop(stream: AsyncStreamWriter) {.async.} =
+  var wstream = cast[TlsStreamWriter](stream)
+  var engine: ptr SslEngineContext
```

Comment:
It's possible to avoid the use of pointers here if the code below the `engine` variable assignment is extracted into a separate proc taking `SslEngineContext` as a parameter. It would be called like this:

```nim
tlsWriteLoopAux(stream, if wstream.kind == TLSStreamKind.Server: wstream.scontext.eng
                        else: wstream.ccontext.eng)
```

The `if` part can be further extracted into a template to make it look even more straight-forward.

Reply:
@zah but all the procedures which will be used requires `ptr SslEngineContext`.

Reply:
I guess you are referring to the procedures from the `BearSSL` C module? 

Nim doesn't have very good guidelines for writing FFI modules, but if a certain C function takes a pointer, you can wrap it in Nim either as a proc taking `ptr` or as a regular Nim proc taking a value type. One advantage of using regular value types is that they make it easier to stick to the safer subset of the language. If we use some kind of static analysis tool that looks for unsafe code, we'll have less flagged code that must be reviewed. 

So, perhaps the BearSSL wrapper can be modified to not user `ptr` parameters, @yglukhov, what do you think?

Reply:
@zah, it can indeed. Current bindings were generated automatically though, and doing nice higher level bindings would require some more manual work. At this point I think it's ok for chronos to use the low level bindings as is.